accord-core/src/test/java/accord/burn/BurnTest.java

```diff
@@ -230,37 +237,50 @@ static void burn(RandomSource random, TopologyFactory topologyFactory, List<Id>
 
         // not used for atomicity, just for encapsulation
         AtomicReference<Runnable> onSubmitted = new AtomicReference<>();
-        Consumer<Packet> responseSink = packet -> {
-            ListResult reply = (ListResult) packet.message;
-            if (replies[(int)packet.replyId] != null)
-                return;
-
+        Runnable maybeTriggerNextRequest = () -> {
```

Comment:
Why is this refactor necessary? The explanation doesn't quite make sense to me, as we were always enqueueing the next message before, unless we had already done so by receiving a later response first. 

I think in both version we potentially breach our concurrency restrictions, as we will enqueue for both the heartbeat and any future response.

Reply:
The issue I faced was that so many txn failed and got delayed to `onDone` that we stoped enqueuing the next packets.  We detected that this happened and would do `return;` and not start the next packets.

> I think in both version we potentially breach our concurrency restrictions, as we will enqueue for both the heartbeat and any future response.

I saw this code as limiting the number of packets and not in-flight `Txn`, so we see a client exception (timeout / fault) and would enqueue the next.

Reply:
> as we will enqueue for both the heartbeat and any future response.

"should" there be any "future responses"?  We send the heart beat IFF the client saw a failure (other than Invalidated) and enqueue trying to figure out the state in `onDone`.  `onDone` is only triggered once we finished processing all the inflight work, which implies the response sink stopped enqueueing work.

The issue I saw was that we stopped enqueuing work, so we fail with `throw new AssertionError("Incomplete set of responses; clock=" + clock.get() + ", expected operations=" + (operations * 2));`.  The root cause I found was that if we don't enqueue on the heart beat there are conditions where we stop progressing the backlog and complete early and fail as the expected amount of operations didn't happen.

Reply:
> "should" there be any "future responses"? We send the heart beat IFF the client saw a failure

Yes, the client will later receive a message if the transaction is recovered - at least to notify of its successful application (I think we do not notify the client on invalidation). 

> I saw this code as limiting the number of packets and not in-flight Txn, so we see a client exception (timeout / fault) and would enqueue the next.

Yes, but the client will (hopefully) receive the later _valid_ response that informs of the outcome of the transaction, at which point it will trigger _another_ pack send, breaching our limits.

> The issue I saw was that we stopped enqueuing work

I don't think that error meant that we had stopped enqueueing work that was still pending - it means we didn't get a complete set of responses to the work we had successfully enqueued. Reaching that message should mean we *must have* successfully enqueued every client packet. We don't stop the simulation until we have sent all packets and processed all ensuing work.

Reply:
Looking at the git history I think this is a by product of writing this patch months ago

```
f7fc6e18aacdeb902f4482ae9fe9429c7e5b8bf8	Fri May 12 01:53:04 PDT 2023	Benedict Elliott Smith	Permit nodes to join a cluster without the full transaction history
```

That patch made sure we notify on all `CoordinationFailure`, so this logic isn't needed anymore

Reply:
reverted this change